log: Symlink privilege issue with Windows

**QUESTION**

In Windows 7 64-bit, I am getting:

```
C:\temp>cockroach.exe start
I170509 11:30:09.103935 1 cli/start.go:575  
CockroachDB OSS v1.0-rc.2 (windows amd64, built 2017/05/05 13:59:27, go1.8.1)
log: exiting because of error: log: 
cannot create log: symlink cockroachexe.xxxxxx.2017-05-09T11_30_09-07_00.001068.log 
C:\temp\cockroach-data\logs\cockroachexe.log: A required privilege is not held by the client.
```

My expectation is that  non-privileged Windows user will be able to run crdb? Is there anyway to avoid this symlink creation?

Well, with the symlink issue gone now ... still no luck - I am running in 64-bit Windows 7, but it is looking for stuff in my kernel32.dll?
```
C:\temp>cockroach.exe start --insecure --log-dir=
I170509 11:58:04.917326 1 cli/start.go:575  CockroachDB OSS v1.0-rc.2 (windows amd64, built 2017/05/05 13:59:27, go1.8.1)
I170509 11:58:05.018336 9 cli/start.go:338  starting cockroach node
W170509 11:58:05.018336 9 server/server.go:162  running in insecure mode, this is strongly discouraged. See --insecure.
E170509 11:58:05.019336 9 util/log/crash_reporting.go:64  a panic has occurred!
panic: Failed to find GetSystemTimePreciseAsFileTime procedure in kernel32.dll: The specified procedure could not be found. [recovered]
        panic: Failed to find GetSystemTimePreciseAsFileTime procedure in kernel32.dll: The specified procedure could not be found.

goroutine 9 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.RecoverAndReportPanic(0x4f628c0, 0xc0424c3f80)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:58 +0x9c
panic(0x137fb80, 0xc0422ffec0)
        /usr/local/go/src/runtime/panic.go:489 +0x2dd
github.com/cockroachdb/cockroach/vendor/golang.org/x/sys/windows.(*LazyProc).mustFind(0xc04237d680)
        /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sys/windows/dll_windows.go:287 +0x5f
github.com/cockroachdb/cockroach/vendor/golang.org/x/sys/windows.(*LazyProc).Addr(0xc04237d680, 0xc0425096e0)
        /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sys/windows/dll_windows.go:294 +0x32
github.com/cockroachdb/cockroach/vendor/golang.org/x/sys/windows.GetSystemTimePreciseAsFileTime(0xc0425096b8)
        /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sys/windows/zsyscall_windows.go:825 +0x34
github.com/cockroachdb/cockroach/pkg/util/timeutil.now(0x7c0349, 0x279bcb4, 0xc0425097f0)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/timeutil/now_windows.go:33 +0x3f
github.com/cockroachdb/cockroach/pkg/util/timeutil.Now(0x0, 0x14bd051fb063ee20, 0x9)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/timeutil/time.go:43 +0x30
github.com/cockroachdb/cockroach/pkg/util/metric.newSlidingHistogram(0xdf8475800, 0x2540be400, 0x1, 0x50)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/metric/sliding_histogram.go:43 +0x42
github.com/cockroachdb/cockroach/pkg/util/metric.NewHistogram(0x154e5de, 0x12, 0x15893bd, 0x35, 0x0, 0x0, 0x0, 0xdf8475800, 0x2540be400, 0x1, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/metric/metric.go:174 +0x53
github.com/cockroachdb/cockroach/pkg/util/metric.NewLatency(0x154e5de, 0x12, 0x15893bd, 0x35, 0x0, 0x0, 0x0, 0xdf8475800, 0x1264f10)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/metric/metric.go:194 +0x74
github.com/cockroachdb/cockroach/pkg/rpc.newRemoteClockMonitor(0xc042273a40, 0x6fc23ac00, 0xdf8475800, 0xc042273ac0)
        /go/src/github.com/cockroachdb/cockroach/pkg/rpc/clock_offset.go:94 +0x3d4
github.com/cockroachdb/cockroach/pkg/rpc.NewContext(0x2220680, 0x27bd388, 0x0, 0xc0421b8d20, 0x4f628c0, 0xc0422ffd70, 0xc0420d42d0, 0xc042273a40, 0xc0422866e0,
0x6600000066)
        /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:170 +0x18f
github.com/cockroachdb/cockroach/pkg/server.NewServer(0xc0420d42d0, 0x0, 0x0, 0xc0422baa00, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:165 +0x474
github.com/cockroachdb/cockroach/pkg/cli.runStart.func1.1(0x4f628c0, 0xc0424c3f80, 0xc0420041d8, 0xc0422866e0, 0xc04214d5b0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/start.go:344 +0x15e
github.com/cockroachdb/cockroach/pkg/cli.runStart.func1(0x4f628c0, 0xc0424c3f80, 0x222c7a0, 0x27bd388, 0xc0420041d8, 0xc0422866e0, 0xc04214d5b0, 0xc0423074a0)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/start.go:414 +0xe6
created by github.com/cockroachdb/cockroach/pkg/cli.runStart
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/start.go:417 +0x3a2
```

--- a/pkg/util/log/file.go
+++ b/pkg/util/log/file.go
@@ -29,6 +29,7 @@ import (
 	"os/user"
 	"path/filepath"
 	"regexp"
+	"runtime"
 	"sort"
 	"strconv"
 	"strings"
@@ -234,8 +235,21 @@ func create(
 	f, err = os.OpenFile(fname, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0664)
 	if err == nil {
 		symlink := filepath.Join(dir, link)
-		_ = os.Remove(symlink) // ignore err
-		err = os.Symlink(filepath.Base(fname), symlink)
+
+		// Symlinks are best-effort.
+
+		if err := os.Remove(symlink); err != nil && !os.IsNotExist(err) {
+			fmt.Fprintf(OrigStderr, "log: failed to remove symlink %s: %s", symlink, err)
+		}
+		if err := os.Symlink(filepath.Base(fname), symlink); err != nil {
+			// On Windows, this will be the common case, as symlink creation
+			// requires special privileges. See https://docs.microsoft.com/en-
+			// us/windows/device-security/security-policy-settings/create-
+			// symbolic-links.
+			if runtime.GOOS != "windows" {
+				fmt.Fprintf(OrigStderr, "log: failed to create symlink %s: %s", symlink, err)
+			}
+		}
 	}
 	return f, updatedRotation, fname, errors.Wrapf(err, "log: cannot create log")
 }

